### PR TITLE
Support for LLaVA

### PR DESCRIPTION
**Pull request by varshith15**
- [x] Single Machine
- [x] Sharding
- [x] Generate function integration
- [x] Clean up code (deduplication)

### CODE REVIEW

**Comment by AlexCheema** [comment]
Looks good so far!

Assigned you on the bounty https://docs.google.com/spreadsheets/d/1cTCpTIp48UnnIvHeLEUNg1iMy_Q6lRybgECSFCoVJpE/edit

**Comment by varshith15** [comment]
hey @AlexCheema sharded inference also works now, integrated it into the generate function as well.
only the main.py integration is left which i feel like should be a seperate PR. So please review and merge.

**Comment by AlexCheema** [comment]
We should merge https://github.com/exo-explore/exo/pull/87 before this as there's some relevant changes.
@varshith15 you can merge that branch already if you want and then I will merge yours as soon as the other is merged.

**Comment by AlexCheema** [comment]
With changes from #87, we should try to re-use as much as possible from `mlx_lm`, there's a lot of copied code when it should only be necessary to copy 1 or 2 classes from there and modify them.

**Comment by AlexCheema** [comment]
Is the `test_sharded_llava.py` test passing for you?
I thought we'd need to load an mlx-converted model, but it looks like this is the original ordinary model from HF `llava-hf/llava-1.5-7b-hf`

**Comment by AlexCheema** [comment]
> Is the `test_sharded_llava.py` test passing for you? I thought we'd need to load an mlx-converted model, but it looks like this is the original ordinary model from HF `llava-hf/llava-1.5-7b-hf`

NVM this works fine :)

File: exo/inference/mlx/models/llava.py [hunk 1]
```diff
@@ -369,22 +370,25 @@ def __call__(
...
-        self.layers = [
-            TransformerBlock(config=config) for _ in range(config.num_hidden_layers)
-        ]
-        self.norm = nn.RMSNorm(config.hidden_size, eps=config.rms_norm_eps)
+        self.layers = []
+        for i in range(self.num_hidden_layers):
+          if self.shard.start_layer <= i <= self.shard.end_layer:
+            self.layers.append(TransformerBlock(config=config))
+          else:
+            self.layers.append(IdentityBlock())
```
Comment (varshith15) [inline exo/inference/mlx/models/llava.py:389]: this isnt needed right @AlexCheema ?

**Comment by varshith15** [comment]
also thanks thanks for pushing the other changes @AlexCheema , they make a lot of sense

File: exo/inference/mlx/models/llava.py (hunk 1 above)
Comment (AlexCheema) [reply exo/inference/mlx/models/llava.py:389]: It is with the new convention, I think

**Comment by AlexCheema** [comment]
It would be great to hook this up to `chatgpt_api.py` do you want to give that a go @varshith15 ?

**Comment by varshith15** [comment]
@AlexCheema done PRM
the logic to use the pixel_values only in the first pass fits in perfectly with the send_prompt, send_tensor idea (really cool idea)
i've verified, its backward compatible as well
only change left is the UI one, I am little occupied so can't pick that up rn

**Comment by AlexCheema** [comment]
> @AlexCheema done PRM the logic to use the pixel_values only in the first pass fits in perfectly with the send_prompt, send_tensor idea (really cool idea) i've verified, its backward compatible as well only change left is the UI one, I am little occupied so can't pick that up rn

Woah! Incredible work. You even figured out how to generate the grpc service. I’m catching a flight now but will be able to look properly once I arrive.

I can fix up a small UI.

email me at alex@exolabs.net for the bounty.

really awesome work, you rock @varshith15

**Comment by AlexCheema** [comment]
Changed a few things, added support for image upload to tiny chat.
Awesome work @varshith15 please email me alex@exolabs.net for the bounty reward!

**Comment by sofinvalery** [comment]
Noob question. how to run this model?
`python3 exo/inference/mlx/models/llava.py
Traceback (most recent call last):
  File "/Users/valeriysofin/Developer/exo/exo/inference/mlx/models/llava.py", line 12, in <module>
    from .base import IdentityBlock
ImportError: attempted relative import with no known parent package`
And will i be able to contribute having just one macbook air m1 with 8gb of ram?

**Pull request merged**